File: _extensions/capsule/capsule.js
// Chargement différé des capsules vidéo (US-19) : l'iframe n'est créée qu'au clic.
(() => {
  const lire = (bouton) => {
    const figure = bouton.closest(".capsule");
    const titre = bouton.querySelector(".capsule-titre")?.textContent?.trim() || "Capsule vidéo";
    const cadre = document.createElement("iframe");
    // youtube-nocookie.com : pas de cookie de suivi tant que la vidéo n'est pas lancée.
    cadre.src = `https://www.youtube-nocookie.com/embed/${encodeURIComponent(bouton.dataset.video)}?autoplay=1&rel=0`;
    cadre.title = titre;
    cadre.loading = "lazy";
    cadre.allow = "accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture";
    cadre.allowFullscreen = true;
    figure.replaceChild(cadre, bouton);
    cadre.focus();
  };

  document.addEventListener("click", (evenement) => {
    const bouton = evenement.target.closest(".capsule-declencheur");
    if (bouton) lire(bouton);
  });
})();

File: _extensions/capsule/capsule.lua
-- Shortcode « capsule » (US-19) : une vidéo YouTube, chargée seulement au clic.
--
--     {{< capsule dQw4w9WgXcQ titre="Séance 1 — l'architecture en couches" >}}
--
-- Tant que le lecteur n'a pas cliqué, **aucune requête ne part vers Google** : la vignette est dessinée
-- par le site, sans appel à img.youtube.com, et l'iframe n'existe pas. Au clic, elle est créée sur
-- youtube-nocookie.com, qui ne dépose pas de cookie de suivi avant lecture.
--
-- Accessibilité : le bouton porte un nom explicite, et l'iframe reçoit le même titre.

local function echapper(texte)
  return (texte:gsub("&", "&amp;"):gsub("<", "&lt;"):gsub(">", "&gt;"):gsub('"', "&quot;"))
end

return {
  ["capsule"] = function(args, kwargs)
    local identifiant = pandoc.utils.stringify(args[1] or "")
    local titre = pandoc.utils.stringify(kwargs["titre"] or "")
    if identifiant == "" then
      error("shortcode capsule : identifiant de la vidéo manquant.")
    end
    if titre == "" then
      error("shortcode capsule : attribut « titre » obligatoire, il nomme la vidéo pour tous les lecteurs.")
    end
    if not identifiant:match("^[%w_-]+$") then
      error("shortcode capsule : identifiant inattendu « " .. identifiant .. " ».")
    end

    quarto.doc.add_html_dependency({ name = "capsule", scripts = { "capsule.js" },
                                     stylesheets = { "capsule.css" } })

    return pandoc.RawBlock("html", string.format([[
<figure class="capsule">
  <button type="button" class="capsule-declencheur" data-video="%s" aria-label="Lire la vidéo : %s">
    <span class="capsule-icone" aria-hidden="true"></span>
    <span class="capsule-titre">%s</span>
    <span class="capsule-note">La vidéo est hébergée par YouTube : rien n'est chargé avant le clic.</span>
  </button>
  <noscript>
    <a href="https://www.youtube.com/watch?v=%s">Voir la vidéo sur YouTube : %s</a>
  </noscript>
</figure>]], echapper(identifiant), echapper(titre), echapper(titre),
               echapper(identifiant), echapper(titre)))
  end,
}
